stage1/init/init.go

```diff
@@ -435,13 +430,32 @@ func getArgsEnv(p *stage1commontypes.Pod, flavor string, canMachinedRegister boo
 		return nil, nil, fmt.Errorf("unrecognized stage1 flavor: %q", flavor)
 	}
 
+	// systemd-nspawn needs /etc/machine-id to link the container's journal
+	// to the host. Since systemd-v230, /etc/machine-id is mandatory, see
+	// https://github.com/systemd/systemd/commit/e01ff70a77e781734e1e73a2238af2e9bf7967a8
+	mPath := filepath.Join(common.Stage1RootfsPath(p.Root), "etc", "machine-id")
 	machineIDBytes := append([]byte(machineID), '\n')
 	if err := ioutil.WriteFile(mPath, machineIDBytes, 0644); err != nil {
 		return nil, nil, errwrap.Wrap(errors.New("error writing /etc/machine-id"), err)
 	}
 	if err := user.ShiftFiles([]string{mPath}, &p.UidRange); err != nil {
 		return nil, nil, errwrap.Wrap(errors.New("error shifting /etc/machine-id"), err)
 	}
+	// also create /etc/machine-id inside each app, so they can read the pod journal when available
```

Comment:
Sure, I can split that into a different PR.

Doing the `stat` before would break some very common cases though. Unfortunately, many common base images have a `/etc/machine-id` baked in as artifacts of how they were built. E.g.:

```
rkt --insecure-options=image run docker://debian:jessie -- -c "cat /etc/machine-id"
[106023.438284] debian[5]: 5d12f049cbc39eb705e94d4b4fe4e580
```